# Our first Kaggle competition, or how to get a decent score without any Machine Learning

---
### by Alexandru Brie

I've been a software developer for over 14 years, and have followed multiple ML/AI tutorials. Still, I consider myself a ML/DS noob. Particularly, although I've read some tutorials on pandas or numpy, I couldn't find any quick and simple walkthrough on how to make an actual submission to Kaggle. 

This tutorial is an entry level tutorial that helps people get over their shyness of joining Kaggle competitions and only makes use of common sense, while gently introducting one into using Pandas and Numpy.

No statistics, Data Science or Machine Learning were used for the making of this tutorial.

What one must know, though, is use Python and Jupyter Notebook and have a couple of Python packages installed on their machine.
The easiest way to do this is by heading to [Anaconda](https://www.anaconda.com/download/) and download the Python 3.7 GUI installer bundle. It will automatically install the most used libraries, including those we're using below. 

Once Anaconda installed, start the "Anaconda Navigator" launcher app, then from it the "Jupyter Notebook". Alternatively, from the command line, type "jupyter notebook". 

This will open a web interface where you can browse, open and create new "notebooks" (a kind of Python programs that let you write and run code, edit text and plot images, in an interactive fashion, and which are used by probably 99% of people in Data Science and Machine Learning). This tutorial was written in such a notebook.


## 1. First, join a Kaggle competition. 

Create a user account on Kaggle, the world's largest online community of people working in AI, Machine Learning and Data Science. 

The Kaggle [Titanic Survivors](https://www.kaggle.com/c/titanic) competition is the one any Kaggle newcomer should start with, as it's always open (leaderboard periodically cleans up), straightforward to follow and easy to understand.

> Knowing from a training set of samples listing passengers who survived or did not survive the Titanic disaster, can our model determine based on a given test dataset not containing the survival information, if these passengers in the test dataset survived or not.

After you join the competition, go to the Data section, click "Download All" and unzip the downloaded archive in a subfolder of where you'll want to write your Jupyter Notebook program.  
Me, I'm working in the "MyWork" folder and saved the downloaded data to the "titanic_data" subfolder.

## Import used libraries

In [1]:
# MUST-USE Python library for handling arrays or matrices
import numpy as np

# Building over numpy, Pandas is the MUST-USE Python library for tabular data
import pandas as pd

# Must-Use plotting/graphing library
import matplotlib.pyplot as plt
%matplotlib inline

## 2. Load training & testing data 

In [2]:
# in my case, the titanic input files are in the "titanic_data" subfolder
PATH = "./titanic_data/"

In [3]:
# check if our input data files are actually at that path
# ! sign calls a system command ; to use a Python variable inside the command, we wrap it in {}

!ls {PATH}

gender_submission.csv test.csv              train.csv


In [4]:
# read and load a csv file into a Pandas "table" (DataFrame), in one line:

# load the train.csv file
train_raw = pd.read_csv(f"{PATH}/train.csv")

# load the test.csv file
test_raw = pd.read_csv(f"{PATH}/test.csv")

Both train_raw and test_raw are Pandas "DataFrame" objects, providing useful methods and variables for manipulating tabular data, accessing any row, column or cell.

We usually access a column using the syntax dataframe["ColumnName"] but occasionally, for brevity, we'll use dataframe.ColumnName instead.

In [5]:
# take a look at the 5 fist lines of the train_raw DataFrame:
train_raw.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S

One awesome feature of data frames is that we can apply selections on them. For instance, here's the list of female passengers who survived:

In [6]:
# when you don't do anything with a dataframe (in our case, the dataframe resulting from a selection)
# Pandas outputs it in a compacted manner, which is useful to get an idea of how both its head and tail look like

train_raw[(train_raw["Sex"]=="female") & (train_raw["Survived"]==1)]

PassengerId  Survived  Pclass  \
1              2         1       1   
2              3         1       3   
3              4         1       1   
8              9         1       3   
9             10         1       2   
10            11         1       3   
11            12         1       1   
15            16         1       2   
19            20         1       3   
22            23         1       3   
25            26         1       3   
28            29         1       3   
31            32         1       1   
32            33         1       3   
39            40         1       3   
43            44         1       2   
44            45         1       3   
47            48         1       3   
52            53         1       1   
53            54         1       2   
56            57         1       2   
58            59         1       2   
61            62         1       1   
66            67         1       2   
68            69         1       3   
79            80         1       3   
82            83         1       3   
84            85         1       2   
85            86         1       3   
88            89         1       1   
..           ...       ...     ...   
774          775         1       2   
777          778         1       3   
779          780         1       1   
780          781         1       3   
781          782         1       1   
786          787         1       3   
796          797         1       1   
797          798         1       3   
801          802         1       2   
809          810         1       1   
820          821         1       1   
823          824         1       3   
829          830         1       1   
830          831         1       3   
835          836         1       1   
842          843         1       1   
849          850         1       1   
853          854         1       1   
855          856         1       3   
856          857         1       1   
858          859         1       3   
862          863         1       1   
865          866         1       2   
866          867         1       2   
871          872         1       1   
874          875         1       2   
875          876         1       3   
879          880         1       1   
880          881         1       2   
887          888         1       1   

                                                  Name     Sex   Age  SibSp  \
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                               Heikkinen, Miss. Laina  female  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
8    Johnson, Mrs. Oscar W (Elisabeth Vilhelmina Berg)  female  27.0      0   
9                  Nasser, Mrs. Nicholas (Adele Achem)  female  14.0      1   
10                     Sandstrom, Miss. Marguerite Rut  female   4.0      1   
11                            Bonnell, Miss. Elizabeth  female  58.0      0   
15                    Hewlett, Mrs. (Mary D Kingcome)   female  55.0      0   
19                             Masselmani, Mrs. Fatima  female   NaN      0   
22                         McGowan, Miss. Anna "Annie"  female  15.0      0   
25   Asplund, Mrs. Carl Oscar (Selma Augusta Emilia...  female  38.0      1   
28                       O'Dwyer, Miss. Ellen "Nellie"  female   NaN      0   
31      Spencer, Mrs. William Augustus (Marie Eugenie)  female   NaN      1   
32                            Glynn, Miss. Mary Agatha  female   NaN      0   
39                         Nicola-Yarred, Miss. Jamila  female  14.0      1   
43            Laroche, Miss. Simonne Marie Anne Andree  female   3.0      1   
44                       Devaney, Miss. Margaret Delia  female  19.0      0   
47                           O'Driscoll, Miss. Bridget  female   NaN      0   
52            Harper, Mrs. Henry Sleeper (Myna Haxtun)  female  49.0      1   
53   Faunthorpe, Mrs. Lizzie (Elizabeth Anne Wilkin...  female  29.0

See details about what the columns in the data file mean, on the [competition data page](https://www.kaggle.com/c/titanic/data).

#### Let's see how Survival is correlated with other data
Pandas crosstab function lets you see how one feature(column) is related to another
It displays a nice table counting values from one on the vertical, and those from the other on the horizontal. 

Since we imported pandas as pd (it's the most common convention), we'll use it this way:

In [7]:
# let's see if women survived more than men
pd.crosstab(train_raw.Survived, train_raw.Sex, margins=True)

Sex       female  male  All
Survived                   
0             81   468  549
1            233   109  342
All          314   577  891

In Pandas, we can address a column(feature) by its name(which, in our csv file, was read from the first line). We'd usually say 

In [8]:
# once more, but with probabilities
pd.crosstab(train_raw.Survived, train_raw.Sex, margins=True, normalize=True)

Sex         female      male       All
Survived                              
0         0.090909  0.525253  0.616162
1         0.261504  0.122334  0.383838
All       0.352413  0.647587  1.000000

Now we can draw primary conclusions, based on the training data file: 

1. 65% of the people in the training data were men(577), 35% were women(314)
2. Out of all men, around 80% died (0.523 / 0.647 , aka 468 / 577)
3. Out of all women, fewer than 26% died (81 / 314)
4. Out of all people, 62% died and 38% survived (the "All" column)

Based on the above insights, we can make several trivial attempts to a solution.

In [9]:
answer_everyone_died = pd.DataFrame({"PassengerId": test_raw["PassengerId"],
                                     "Survived": 0})
answer_everyone_died.to_csv(f"{PATH}/submission_zeros.csv", index=False)

Now, we head to the Kaggle Titanic competition page, and click the "Submit Prediction" button on the tabs menu. Drag and drop the "submission_zeros.csv" file that we've generated above, into the "Upload Files" area, then click the "Make Submission" button.

Wait for the score to be computed and...

We got a 0.72727 accuracy score, which puts us at around #9000 out of over #10000 competitors (at the date of writing this, November 16 2018). This means that over 1000 people were more wrong than us - that is more wrong than the most trivial answer.

## 2. Gender submission
How about saying that all men died, and all women survived? Kaggle already helps us with this, and they provide a gender_submission.csv file inside their data archive. You can upload it directly, for the heck of it, or we can generate one ourselves, in the following cell:

In [10]:
# make a copy of the test data, so as not to mess with it
test_gender_submission = test_raw.copy()

# add a new column that's by default zero
test_gender_submission["Survived"] = 0

# for everyone with the Sex value "female", set the value on the "Survived" column to 1
test_gender_submission.loc[(test_gender_submission["Sex"] == "female"), "Survived"] = 1

# now let's create a new DataFrame with only the answer and save it 
my_gender_submission = pd.DataFrame({"PassengerId": test_gender_submission["PassengerId"],
                                     "Survived": test_gender_submission["Survived"]})
my_gender_submission.to_csv(f"{PATH}/my_gender_submission.csv", index=False)

We can test for differences between our version and the one we got from Kaggle
> diff -w my_gender_submission.csv gender_submission.csv

And they are identical (with the possible exception of whitespace, depending on your computer/operating system).

Uploading the new file to Kaggle and testing how it scores gives us a score of 0.76555 which puts us somewhere around #6300 out of over #10000 competitors.

In other words, almost 4000 people (40%) didn't do any better than just upload the default answer they got offered from Kaggle.

Our basic tutorial can probably stop here, as it already achieved its purpose, showing you a lot of things: 

* how to join a Kaggle competition
* how to start writing code to give an answer to the competition
* how to load data in csv files
* how to look at the first 5 rows of the data (the head method)
* how to use the crosstab command to see how one feature relates to another
* how to create a new DataFrame for the solution and save it to a csv file
* how to submit your solution file to Kaggle and learn your score

But, our no-coding "gender solution" placed us in the last 40%. This makes me curious if, with very little coding or knowledge, we can get a bit further up the leaderboard. 
We'll need to use a couple more methods from the pandas/numpy libraries.

# 3. Tweaked gender solution

We've gotten a reasonable score by claiming that all men died and all women survived. This is obviously not 100% true. Let's see if we can do better. 

We've seen in movies that usually, when a ship is sinking, women and children are usually put first on the boats. Also, in that Titanic movie, it seemed that rich people usually survived while the poor ones didn't. Let's check these assumptions using the tools we have.

I'll add a new column that says if a passenger was a woman or a child (10 years or less) and call it "MightSurvive" (to be used for other survival criteria as well)

In [11]:
# make a copy so as not to change the input data
train_2 = train_raw.copy()
train_2["MightSurvive"] = 0
train_2.loc[(train_2["Sex"]=="female") | (train_2["Age"] < 10), "MightSurvive"] = 1

In [12]:
# let's see how it looks
train_2.head(10)

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   
5            6         0       3   
6            7         0       1   
7            8         0       3   
8            9         1       3   
9           10         1       2   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   
5                                   Moran, Mr. James    male   NaN      0   
6                            McCarthy, Mr. Timothy J    male  54.0      0   
7                     Palsson, Master. Gosta Leonard    male   2.0      3   
8  Johnson, Mrs. Oscar W (Elisabeth Vilhelmina Berg)  female  27.0      0   
9                Nasser, Mrs. Nicholas (Adele Achem)  female  14.0      1   

   Parch            Ticket     Fare Cabin Embarked  MightSurvive  
0      0         A/5 21171   7.2500   NaN        S             0  
1      0          PC 17599  71.2833   C85        C             1  
2      0  STON/O2. 3101282   7.9250   NaN        S             1  
3      0            113803  53.1000  C123        S             1  
4      0            373450   8.0500   NaN        S             0  
5      0            330877   8.4583   NaN        Q             0  
6      0             17463  51.8625   E46        S             0  
7      1            349909  21.0750   NaN        S             1  
8      2            347742  11.1333   NaN        S             1  
9      0            237736  30.0708   NaN        C             1

In [13]:
pd.crosstab(train_2["Survived"], train_2["MightSurvive"], margins=True)

MightSurvive    0    1  All
Survived                   
0             455   94  549
1              90  252  342
All           545  346  891

So, how does this compare with that time when we claimed that only women survived? 

Looking again at the crosstab we made after loading the training data we notice that, when we said that only women survived and all men died, we were right 233(women who actually survived) + 468(men who actually died) out of a total of 891 times. That is circa 78.6% times right.

Now, saying that only women OR children survived, we'd be right 252 + 455 times out of 891. That is circa 79.3%, a small but significant improvement.

Let's submit a this solution and see how it scores.

In [14]:
# make a copy of the test data, so as not to mess with it
test_gender_children_submission = test_raw.copy()

# add the MightSurvive column that's by default zero
test_gender_children_submission["MightSurvive"] = 0

# women and children are likely to survive
test_gender_children_submission.loc[(test_gender_children_submission["Sex"]=="female") |
                                    (test_gender_children_submission["Age"] < 10), "MightSurvive"] = 1

# now let's create a new DataFrame with only the answer and save it 
gender_children_submission = pd.DataFrame({"PassengerId": test_gender_children_submission["PassengerId"],
                                     "Survived": test_gender_children_submission["MightSurvive"]})
gender_children_submission.to_csv(f"{PATH}/gender_children_submission.csv", index=False)

Kaggle grades us with a 0.77033 score. Hurray, this is quite the accomplishment. We've moved up to around #5500 of the #10100 leaderboard, that is in the top 55%.

### What about that "rich people survived" claim?
The passenger wealth is correlated to the passenger class (first class is better than 3rd class). Let's see if this helps.

In [15]:
pd.crosstab(train_2["Survived"], train_2["Pclass"], margins=True)

Pclass      1    2    3  All
Survived                    
0          80   97  372  549
1         136   87  119  342
All       216  184  491  891

Saying that only people in Pclass 1 survived and everyone else died would be right 136(actual class 1 survivors) + 469 (97 actual non-survivors in class 2 + 372 actual non-survivor in class 3) out of 891 times, that is  only 67% times.

That's worse than the above estimations. Maybe we shouldn't focus so much on passenger class for our manual attempt.

### What about small children in large families?

I'd like to check the hypothesis that members of large families were more likely to die. This might be anecdotical, but I'd imagine poor immigrant large families struggling to gather everyone left behind and thus failing to get to the rescue boats. Maybe their parents survived, but most children probably not.
This might not be a general truth, but a couple of families in this situation might give us a better score already. 

The SibSp (Siblings and Spouse) and Parch(Parents and Children) columns cover this information. A crosstab shows how it is, indeed, for large numbers, correlated with non-survival.

In [16]:
pd.crosstab(train_2.Survived, train_2.SibSp, margins=True)

SibSp       0    1   2   3   4  5  8  All
Survived                                 
0         398   97  15  12  15  5  7  549
1         210  112  13   4   3  0  0  342
All       608  209  28  16  18  5  7  891

Looks like children with at least 4 siblings were unlikely to survive.

In [17]:
pd.crosstab(train_2.Survived, train_2.Parch, margins=True)

Parch       0    1   2  3  4  5  6  All
Survived                               
0         445   53  40  2  4  4  1  549
1         233   65  40  3  0  1  0  342
All       678  118  80  5  4  5  1  891

Looks like parents with at least 4 children were unlikely to survive.

Let's add the above findings to our MightSurvive criteria (which is, so far, made of women or children)

In [18]:
train_2[(train_2.Parch >= 4) & (train_2.MightSurvive == 1)].Survived.value_counts()

0    6
1    1
Name: Survived, dtype: int64

In [19]:
train_2[(train_2.SibSp >= 4) & (train_2.MightSurvive == 1)].Survived.value_counts()

0    19
1     3
Name: Survived, dtype: int64

In [20]:
train_2.loc[(train_2.Parch >= 4), "MightSurvive"] = 0

In [21]:
train_2.loc[(train_2.SibSp >= 4), "MightSurvive"] = 0

In [22]:
pd.crosstab(train_2["Survived"], train_2["MightSurvive"], margins=True)

MightSurvive    0    1  All
Survived                   
0             480   69  549
1              94  248  342
All           574  317  891

Our crosstab says that our "MightSurvive" criteria would be right (248 + 480)/891 = 81.7% . That's significantly better than before, almost 3% improvement over the gender solution.

### Let's build our solution and send it to Kaggle for scoring.

In [23]:
# make a copy of the test data, so as not to mess with it
test_tweaked_submission = test_raw.copy()

# add the MightSurvive column that's by default zero
test_tweaked_submission["MightSurvive"] = 0

# women and children are likely to survive
test_tweaked_submission.loc[(test_tweaked_submission["Sex"]=="female") | 
                            (test_tweaked_submission["Age"] < 10), "MightSurvive"] = 1
# mothers with 4+ children were unlikely to survive
test_tweaked_submission.loc[(test_tweaked_submission["Parch"] >= 4), "MightSurvive"] = 0
# children with 4+ siblings were unlikely to survive
test_tweaked_submission.loc[(test_tweaked_submission["SibSp"] >= 4), "MightSurvive"] = 0


# now let's create a new DataFrame with only the answer and save it 
tweaked_submission = pd.DataFrame({"PassengerId": test_tweaked_submission["PassengerId"],
                                     "Survived": test_tweaked_submission["MightSurvive"]})
tweaked_submission.to_csv(f"{PATH}/manually_tweaked_submission.csv", index=False)

Now we're talking, as the score from Kaggle is 0.78468 - better than before.

This puts us around #3200 in the leaderboard. We're in the top 32%, with no machine learning or statistics used so far.

#### Congratulations!